base/inference.jl

```diff
@@ -736,6 +747,174 @@ function limit_type_depth(t::ANY, d::Int, cov::Bool, vars::Vector{TypeVar}=TypeV
+    if isa(c, TypeVar)
+        if isa(t, TypeVar)
+            return type_more_complex(t.ub, c.ub, sources) ||
+                   type_more_complex(t.lb, c.lb, sources);
+        end
+    elseif isa(c, Union)
+        if isa(t, Union)
+            return type_more_complex(t.a, c.a, sources) ||
+                   type_more_complex(t.b, c.b, sources);
+        end
+        return type_more_complex(t, c.a, sources) &&
+               type_more_complex(t, c.b, sources);
+    elseif isa(c, UnionAll)
+        if isa(t, UnionAll)
+            return type_more_complex(t.body, c.body, sources);
+        end
+    elseif isa(t, DataType)
+        if isa(c, DataType)
+            tP = t.parameters
+            isempty(tP) && return false
+            if t.name === c.name
+                cP = c.parameters
+                length(tP) > length(cP) && return true
+                for i = 1:length(tP)
+                    type_more_complex(tP[i], cP[i], sources) && return true
+                end
+            else # allow deviation in the type name, if all of it's parameters came from the sources
+                for i = 1:length(tP)
+                    for s in sources
+                        is_derived_type(tP[i], s) || return true
+                    end
+                end
+            end
+            return false
+        end
+        if isType(t) # allow taking typeof a source type as Type{...}, as long as it isn't nesting Type{Type{...}}
+            tt = unwrap_unionall(t.parameters[1])
+            if isa(tt, DataType) && !isType(tt)
+                for s in sources
+                    is_derived_type(tt, s) && return false
+                end
+            end
+        end
+    elseif t === c
+        return false
+    elseif isa(t, Int) && isa(c, Int)
+        return t === 1 # alternatively: 0 <= t < c
+    end
+    return true
+end
+
+function is_derived_type(t::ANY, c::ANY) # try to find `type` somewhere in `comparison` type
+    t === c && return true
+    if isa(c, Union)
+        return is_derived_type(t, c.a) || is_derived_type(t, c.b)
+    elseif isa(c, UnionAll)
+        return is_derived_type(t, c.body)
+    elseif isa(c, DataType)
+        super = supertype(c)
+        while super !== Any
+            t === super && return true
+            super = supertype(super)
+        end
+        cP = c.parameters
+        for p in cP
+            is_derived_type(t, p) && return true
+        end
+    end
+    return false
+end
+
+function _limit_type_size(t::ANY, c::ANY, sources::SimpleVector) # type vs. comparison which was derived from source
+    t === c && return t # quick egal test
+    isa(t, Type) && isa(c, Type) && c <: t && return t # t is already wider than the comparison
+    let baset = unwrap_unionall(t)
+        for s in sources
+            is_derived_type(baset, s) && return t # t isn't something new
+        end
+    end
+    if isa(t, TypeVar)
+        if isa(c, TypeVar)
+            if t.ub === c.ub && t.lb === c.ub
```

Comment:
Is it intentional that this is comparing a lower bound to an upper bound, when everywhere else upper is compared to upper and lower is compared to lower? If so, could really use a comment explaining why.
If not intentional and this is a typo, if it passes tests then that's indicative of missing coverage.